File: app/src/main/java/com/judas/katachi/utils/log/Logger.java
package com.judas.katachi.utils.log;

import android.util.Log;

import com.judas.katachi.BuildConfig;

public final class Logger {
    public enum Level {
        // Order by importance (needed for logger filter)
        VERBOSE, DEBUG, INFO, WARNING, ERROR
    }

    private static String PREFIX = "";
    private static final Level LEVEL = BuildConfig.DEBUG ? Level.VERBOSE : Level.WARNING;

    private Logger() {
    }

    public static void setAppName(final String appName) {
        PREFIX = "[" + appName + "] ";
    }

    public static void log(final Level level, final String tag, final String msg) {
        final String message = PREFIX + msg;

        if (level.ordinal() >= LEVEL.ordinal()) {
            switch (level) {
                case VERBOSE:
                    Log.v(tag, message);
                    break;
                case DEBUG:
                    Log.d(tag, message);
                    break;
                case INFO:
                    Log.i(tag, message);
                    break;
                case WARNING:
                    Log.w(tag, message);
                    break;
                case ERROR:
                    Log.e(tag, message);
                    break;
                default:
                    break;
            }
        }
    }

    public static void logWithTrace(final Level level, final String tag, final String msg) {
        if (level.ordinal() >= LEVEL.ordinal()) {

            final String log = msg + Log.getStackTraceString(new Exception("logWithTrace"));

            switch (level) {
                case VERBOSE:
                    Log.v(tag, log);
                    break;
                case DEBUG:
                    Log.d(tag, log);
                    break;
                case INFO:
                    Log.i(tag, log);
                    break;
                case WARNING:
                    Log.w(tag, log);
                    break;
                case ERROR:
                    Log.e(tag, log);
                    break;
                default:
                    break;
            }
        }
    }

    public static void logError(final String tag, final String target, final Throwable error) {
        log(Level.ERROR, tag, target + " ERROR: " + error.getMessage());
    }
}

File: app/src/main/kotlin/com/judas/katachi/utility/Logger.kt
package com.judas.katachi.utility

import android.util.Log

object Logger {
    enum class Level { VERBOSE, DEBUG, INFO, WARNING, ERROR }

    var logLevel: Level = Level.ERROR
    var appName: String = ""

    // Logs a message with Android default logger.
    internal fun log(
        level: Level = Level.VERBOSE,
        tag: String = "",
        message: String = "",
        error: Throwable? = null,
    ) {
        val fullTag = tag.take(23) // Max tag length is 23 on older Android versions

        // Add prefix, error cause & stacktrace if necessary + cut to max length
        var fullMessage = "[$appName] $message"
        error?.let {
            fullMessage += ": ${it.message}. Caused by ${it.cause}\n"
            fullMessage += error.stackTraceToString()
        }

        // Log to Android Logcat
        if (level.ordinal >= logLevel.ordinal) {
            when (level) {
                Level.VERBOSE -> Log.v(fullTag, fullMessage)
                Level.DEBUG -> Log.d(fullTag, fullMessage)
                Level.INFO -> Log.i(fullTag, fullMessage)
                Level.WARNING -> Log.w(fullTag, fullMessage)
                Level.ERROR -> Log.e(fullTag, fullMessage)
            }
        }
    }
}

fun Any.log(
    level: Logger.Level = Logger.Level.VERBOSE,
    message: String = "",
    error: Throwable? = null
) = Logger.log(
    level = level,
    tag = this::class.java.simpleName,
    message = message,
    error = error
)
